Implement PSD write support

ajjahn opened the pull request:
A work in progress, looking for feedback. Among other things, the layer and mask information is still incomplete. 

It has become apparent that in order for writing PSDs to be useful, an API for modifying `Psd` needs to be designed and implemented.

ajjahn committed:
partial write support in progress

mexicantexan commented:
want to collab on this?

ajjahn commented:
> want to collab on this?

That would be ideal. I think we need to start a conversation about the API for creating/updating a `Psd` object in order to write it out as a .psd file. I think that might impact how the internals are represented due to the fact that up until now this library has only read bytes in. We probably need @chinedufn to jump into that conversation too.

My particular use case is to transform a proprietary image composition document type into PSD. That will mean needing to create pixel layers, text objects, set blending modes, using adjustment layers, group and layer masks, even potentially using effects.

mexicantexan commented:
Sounds good to me. My use case is much simpler, I just need to be able to load a standard image type into a layer and save it. So I believe it would get wrapped up into your spec list.

chinedufn commented:
I haven't looked at this PR yet, but if it doesn't contain any tests then that would be something to think about since we don't want to land any untested functionality.

---

We can follow the approach laid out in https://github.com/chinedufn/psd/issues/61#issuecomment-2328520087 until we have enough implemented to support both of your needs.

I'll make myself available to answer questions and review code.

Commit:
Add changes to update dependencies and fix compilation warnings.

Commit:
Update README to fix CircleCI broken link.

mexicantexan committed:
clean build warnings

mexicantexan committed:
Reduce redundant code in sixteen_to_eight_rgba

mexicantexan committed:
Optimize blend_pixels function by ~20%

mexicantexan committed:
Fix edge cases where 'color_burn' and 'color_dodge' would result in incorrect values through the use of epsilon comparisons.
Add tests for color burn and color dodge that explicitly test for this.

mexicantexan committed:
Add blending implementations for 'darker_color', 'lighter_color', 'vivid_light', 'linear_light', 'pin_light', and 'hard_mix'.
Add luminance calculation. Add rgb_to_hsl conversion and hsl_to_rgb conversion. Add hue, saturation, color, and luminosity blending ability.

mexicantexan committed:
Update TODO and blend_pixels docs

mexicantexan commented:
@ajjahn I took some time last night to read through your progress so far and implemented some missing features (specifically around blending) that I think are going to be necessary for us.

mexicantexan committed:
Add ability to write color mode data in color_mode_data_section.rs. Add a simple test to check functionality and backwards compatibility.

mexicantexan committed:
ADD in TODOs
ADD helper functions
ADD helper function comments for clarity

mexicantexan committed:
BUG FIX for wasm-pack in build step

mexicantexan committed:
BUG FIX for wasm-pack error in build step

mexicantexan committed:
BUG FIX for wasm-pack error in build step

mexicantexan committed:
BUG FIX update wasm-pack in circleci build step

mexicantexan committed:
BUG FIX update wasm-pack in circleci build step

mexicantexan committed:
Add RLE-compressed scanline support, enhance compression handling, and implement basic roundtrip test.

mexicantexan commented:
@ajjahn @chinedufn basic round-tripping (read -> save -> read) seems to be functioning

mexicantexan commented:
@chinedufn what would you need from me to merge this PR?

ajjahn commented:
I think we can at least mark this PR as ready for review.

mexicantexan commented:
@chinedufn Just following up on this

chinedufn commented:
Planning to review next week.

Main thing I'll be looking for:

- all of the functionality is tested
  - every test should include a doc comment that explains what is is verifying.
    - `/// Verify that ...`
- the new types/functions/methods/ are well documented

I'm noting this in case you want to look it over first. We will not merge any untested functionality.

Otherwise, I'm planning to review this next week.

Thank you for your efforts here.

mexicantexan committed:
Improve test descriptions and comments for clarity and consistency across modules. Refactor buffer write methods in `sections/mod.rs` to enhance error handling.
